infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/EnvironmentVariableSecretApplier.java

```diff
@@ -0,0 +1,112 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.model.impl.devfile.ComponentImpl;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodRole;
+
+/**
+ * Mounts Kubernetes secret with specific annotations as an environment variable(s) in workspace
+ * containers. Allows per-component control of secret applying in devfile.
+ */
+@Beta
+@Singleton
+public class EnvironmentVariableSecretApplier
+    extends KubernetesSecretApplier<KubernetesEnvironment> {
+
+  static final String ANNOTATION_ENV_NAME = ANNOTATION_PREFIX + "/" + "env-name";
+  static final String ANNOTATION_ENV_NAME_TEMPLATE = ANNOTATION_PREFIX + "/%s_" + "env-name";
+
+  @Override
+  public void applySecret(KubernetesEnvironment env, Secret secret) throws InfrastructureException {
+    boolean secretAutomount =
+        Boolean.parseBoolean(secret.getMetadata().getAnnotations().get(ANNOTATION_AUTOMOUNT));
+    for (PodData podData : env.getPodsData().values()) {
+      if (!podData.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      for (Container container : podData.getSpec().getContainers()) {
+        Optional<ComponentImpl> component = getComponent(env, container.getName());
+        if ((component.isPresent() && isComponentAutomountFalse(component.get()))
+            || (!secretAutomount
+                && !(component.isPresent() && isComponentAutomountTrue(component.get())))) {
+          continue;
+        }
+        for (Entry<String, String> secretDataEntry : secret.getData().entrySet()) {
+          final String mountEnvName = envName(secret, secretDataEntry.getKey());
+          container
+              .getEnv()
+              .add(
+                  new EnvVarBuilder()
+                      .withName(mountEnvName)
+                      .withValueFrom(
+                          new EnvVarSourceBuilder()
+                              .withSecretKeyRef(
+                                  new SecretKeySelectorBuilder()
+                                      .withName(secret.getMetadata().getName())
+                                      .withKey(secretDataEntry.getKey())
+                                      .build())
+                              .build())
+                      .build());
+        }
+      }
+    }
+  }
+
+  private String envName(Secret secret, String key) throws InfrastructureException {
+    String mountEnvName;
+    if (secret.getData().size() == 1) {
+      mountEnvName =
+          Stream.of(
+                  secret.getMetadata().getAnnotations().get(ANNOTATION_ENV_NAME),
```

Comment:
what if secret has both env_name and template env_name annotation?

Reply:
It will pick up first. Generally, i want to avoid to fail on each and every little error, because one bad secret can block all workspaces to start if thet all share the same NS.

Reply:
can we log it then? preferably so it's visible workspace-loader log

Reply:
Reworked the code, can you suggest how to log it properly to be visible on user-s console ?

Reply:
https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/util/RuntimeEventsPublisher.java#L99

Reply:
btw I would log it in che-server log as well. It might be useful for us in bug investigations

Reply:
Added